### PR TITLE
Evaluation of reason expression of `revert` and `require` expression when `--revert-strings strip` is used.

### PR DESCRIPTION
**Pull request by rodiazet**
## Description

This PR fixes inconsistency between legacy and via-ir pipeline regarding evaluation of the reason expression for `require` and `revert` expression in case of `--revert-strings strip` usage.

When `--revert-strings strip` flag is passed and in the legacy pipeline, evaluation of the reason expression in `require`/`revert` functions is omitted when the `isPure` flag for the expression is `true` (mean it’s a runtime constant). It can be `true` also for expressions which can panic. Not sure it was a reason to introduce this additional optimisation (to not evaluate the expression), but in via-ir it’s evaluated and popped from the stack. Legacy pipeline now copies this behaviour. 

## AI Disclosure

- [x] No AI tools were used

### CODE REVIEW

**Review by blishko — COMMENTED** [review]
Can you add also the test case that @cameel suggested? With arithmetic overflow for constants?

**Review by blishko — DISMISSED** [review]
Looks good!

Can you elaborate a bit more in the PR description?
What was the inconsistency between the pipelines? How is this fixing the inconsistency? My intuition is that the reason is always going to be evaluated in the legacy pipeline, so that was the behaviour in IR before as well?

**Review by r0qs — COMMENTED** [review]
Looks good to me. I think this deserves a changelog entry though, since code that previously succeeded can now panic, right?

File: test/libsolidity/SolidityCompiler.cpp [hunk 1]
```diff
@@ -71,4 +124,121 @@ BOOST_AUTO_TEST_CASE(does_not_include_creation_time_only_internal_functions)
 
 BOOST_AUTO_TEST_SUITE_END()
 
+class RevertStringsStripFixture: protected AnalysisFramework
```
Comment (blishko) [inline test/libsolidity/SolidityCompiler.cpp:127]: I think we also need some sanity checks with cases where `assemblyContainsString` returns `true`.

Right now I could change the implementation to just always return `false` and your tests would still be passing.

File: test/libsolidity/semanticTests/revertStrings/reason_expression_evaluation_with_stripped_revert_strings.sol [hunk 2]
```diff
@@ -0,0 +1,139 @@
+contract StripReasonPanic {
+    enum MyEnum { A, B }
+    string constant ARRAY_OUT_OF_BOUNDS_REASON = (new bytes(0))[0] == bytes1(0) ? "first" : "second";
+    string constant OVERFLOW_REASON = type(uint8).max + 1 == 0 ? "first" : "second";
+    string constant DIVISION_BY_ZERO_REASON = 1 / type(uint8).min == 0 ? "first" : "second";
+    string constant ENUM_CONVERSION_REASON = MyEnum(type(uint8).max) == MyEnum.A ? "first" : "second";
+
+    function requireReasonArrayOutOfBoundsIsEvaluated() external returns (uint256) {
+        require(true, (new bytes(0))[0] == bytes1(0) ? "first" : "second");
```
Comment (blishko) [inline test/libsolidity/semanticTests/revertStrings/reason_expression_evaluation_with_stripped_revert_strings.sol:9]: I would change some of these tests (or add new ones) with `require(false, ...)`.
With `require(true, ...)` one would expect the reason to be evaluated. It is not so obvious that this should be the case also when the condition is `false`.

**Review by blishko — CHANGES_REQUESTED** [review]
